src/libs/decoders/opus.c

```diff
@@ -0,0 +1,662 @@
+		                                          OPUS_SAMPLE_RATE,
+		                                          sample->desired.rate,
+		                                          // SPEEX_RESAMPLER_QUALITY_VOIP,    // consumes ~20 Mhz
+		                                          SPEEX_RESAMPLER_QUALITY_DEFAULT,    // consumes ~40 Mhz
+		                                          // SPEEX_RESAMPLER_QUALITY_DESKTOP, // consumes ~80 Mhz
+		                                          &rcode);
+
+		// If we failed to initialize the resampler, then tear down
+		if (rcode < 0) {
+			opus_close(sample);
+			BAIL_MACRO("Opus: failed initializing the resampler", 0);
+		}
+
+	// Otherwise use native sampling
+	} else {
+		sample->actual.rate = OPUS_SAMPLE_RATE;
+		decoder->rate_ratio = 1.0;
+		decoder->resampler = NULL;
+	}
+
+	// Allocate our buffer to hold PCM samples from the Opus decoder
+	decoder->buffer_size = (Uint16) (oh->channel_count * OPUS_MIN_BUFFER_SAMPLES_PER_CHANNEL * 1.5);
+	decoder->buffer = SDL_malloc(decoder->buffer_size * sizeof(opus_int16));
+
+	// Gather static properties about our stream (channels, seek-ability, format, and duration)
+	sample->actual.channels = (Uint8)(oh->channel_count);
+	sample->flags = op_seekable(of) ? SOUND_SAMPLEFLAG_CANSEEK: 0;
+	sample->actual.format = AUDIO_S16LSB; // returns least-significant-byte order regardless of architecture
+
+	ogg_int64_t total_time = op_pcm_total(of, -1); // total PCM samples in the stream
+	internal->total_time = total_time == OP_EINVAL ? -1 : // total milliseconds in the stream
+		                                 (Sint32)( (double)total_time / OPUS_SAMPLE_RATE_PER_MS);
+
+	return 1;
+} /* opus_open */
+
+
+/*
+ * Opus Close
+ * ==========
+ * Free and NULL all allocated memory pointers.
+ */
+static void opus_close(Sound_Sample* sample)
+{
+	/* From the Opus docs: if opening a stream/file/or using op_test_callbacks() fails
+	 * then we are still responsible for freeing the OggOpusFile with op_free().
+	 */
+	Sound_SampleInternal* internal = (Sound_SampleInternal*) sample->opaque;
+
+	opus_t* d = internal->decoder_private;
+	if (d != NULL) {
+
+		if (d->of != NULL) {
+			op_free(d->of);
+			d->of = NULL;
+		}
+
+		if(d->resampler != NULL) {
+			speex_resampler_destroy(d->resampler);
+			d->resampler = NULL;
+		}
+
+		if(d->buffer != NULL) {
+			SDL_free(d->buffer);
+			d->buffer = NULL;
+		}
+
+		SDL_free(d);
+		d = NULL;
+	}
+	return;
+
+} /* opus_close */
+
+
+/*
+ * Opus Read
+ * =========
+ * Decode, resample (if needed), and write the output to the
+ * requested buffer.
+ */
+static Uint32 opus_read(Sound_Sample* sample)
+{
+	Sound_SampleInternal* internal = (Sound_SampleInternal*) sample->opaque;
+	opus_t* d = internal->decoder_private;
+
+	opus_int16* output_buffer = internal->buffer;
+	const Uint16 requested_output_size = internal->buffer_size / sizeof(opus_int16);
+	const Uint16 derived_consumption_size = (Uint16) ((requested_output_size * d->rate_ratio) + 0.5);
```

Comment:
Maybe `ceilf` is more appropriate here?

Reply:
Absolutely; thanks.

Reply:
Fixed in https://github.com/dreamer/dosbox-staging/commit/5f71f3a6f1225306ff7fb7bccd96fd5e315db5df